### PR TITLE
raftstore: report has_commit_merge in unsafe recovery

### LINKED ISSUE #12627: Handle unapplied commit merges in unavailable regions during unsafe recovery.
## Development Task
This is part of #10483 

|| store 1 | store 2| store 3|
|---|---|---|---|
| Initial | A=[a,c) B=[c,z) | A=[a,c) B=[c,z) | A=[a,c) B=[c,z) |
| PrepareMerge | un-replicated | A=[a,c)->target B=[c,z)->source | A=[a,c)->target B=[c,z)->source |
|CommitMerge | replicated uncommitted | A=[a,z) | A=[a,z) |

If the source region enters force leader state before the target, data loss may happen, we need to handle this correctly.

### PR DESCRIPTION
**Pull request by v01dstar**
Signed-off-by: v01dstar <yang.zhang@pingcap.com>

### What is changed and how it works?

Issue Number: Close #12627

What's Changed:

```commit-message
Include `has_commit_merge` info in the store report
```

### Related changes

- N/A

### Check List 

Tests 

- Integration test
Side effects

- Performance regression
N/A

- Breaking backward compatibility
N/A

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -899,6 +899,21 @@ where
...
+        if let Ok(entries) = self.fsm.peer.get_store().entries(
+            self.fsm.peer.raft_group.store().commit_index(),
+            self.fsm.peer.get_store().last_index() + 1,
+            NO_LIMIT,
+            GetEntriesContext::empty(false),
+        ) {
+            for entry in entries {
+                let index = entry.get_index();
+                let cmd: RaftCmdRequest =
+                    util::parse_data_at(entry.get_context(), index, "Unsafe recovery, whether has commit merge");
```
Comment (Connor1996) [inline components/raftstore/src/store/fsm/peer.rs:911]: we can leverage `ProposalContext`, then no need to parse the data

Comment (v01dstar) [reply components/raftstore/src/store/fsm/peer.rs]: Done

File: components/raftstore/src/store/fsm/peer.rs [hunk 2]
```diff
@@ -899,6 +899,17 @@ where
         region_local_state.set_region(self.region().clone());
         self_report.set_region_state(region_local_state);
         self_report.set_is_force_leader(self.fsm.peer.force_leader.is_some());
+        if let Ok(entries) = self.fsm.peer.get_store().entries(
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:902]: What if error is returned?

Comment (v01dstar) [reply components/raftstore/src/store/fsm/peer.rs]: Should panic, added.

**Review by Connor1996 — APPROVED** [review]
LGTM

File: tests/integrations/raftstore/test_unsafe_recovery.rs [hunk 3]
```diff
@@ -1055,3 +1055,71 @@ fn test_force_leader_multiple_election_rounds() {
...
+fn test_unsafe_recovery_has_commit_merge() {
+    let mut cluster = new_node_cluster(0, 3);
+    configure_for_merge(&mut cluster);
+
+    cluster.run();
+
+    cluster.must_put(b"k1", b"v1");
+    cluster.must_put(b"k3", b"v3");
+    let pd_client = Arc::clone(&cluster.pd_client);
+    // pd_client.disable_default_operator();
```
[the remark this answers is not in the document]
Comment (v01dstar) [reply tests/integrations/raftstore/test_unsafe_recovery.rs]: Done

File: components/raftstore/src/store/fsm/peer.rs [hunk 4]
```diff
@@ -899,6 +899,20 @@ where
...
+        match self.fsm.peer.get_store().entries(
+            self.fsm.peer.raft_group.store().commit_index() + 1,
+            self.fsm.peer.get_store().last_index() + 1,
+            NO_LIMIT,
+            GetEntriesContext::empty(false),
+        ) {
+            Ok(entries) => {
+                for entry in entries {
+                    let ctx = ProposalContext::from_bytes(&entry.context);
+                    self_report.set_has_commit_merge(ctx.contains(ProposalContext::COMMIT_MERGE));
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:911]: What if the commit merge entry is in the middle?

Comment (v01dstar) [reply components/raftstore/src/store/fsm/peer.rs]: Sorry, I was being dumb 😢 Updated.

**Review by BusyJay — APPROVED** [review]

**Pull request merged**